ambry-router/src/main/java/com/github/ambry/router/CompressionService.java

```diff
@@ -363,89 +349,90 @@ public ByteBuf decompress(ByteBuf compressedBuffer, int fullChunkSize) throws Co
       throw new CompressionException("Decompression failed due to unknown algorithm name " + algorithmName);
     }
 
-    // Apply decompression, then wrap result in ByteBuf.
-    byte[] decompressedBuffer;
+    // Apply decompression.
     CompressionMetrics.AlgorithmMetrics algorithmMetrics = compressionMetrics.getAlgorithmMetrics(algorithmName);
-    try {
       algorithmMetrics.decompressRate.mark();
-      long startTime = System.nanoTime();
-      decompressedBuffer = decompressor.decompress(buffer, bufferOffset, dataSize);
-      long durationMicroseconds = (System.nanoTime() - startTime)/1000;
 
-      long speedInMBPerSec = durationMicroseconds == 0 ? 0 :
-          (long) (BytePerMicrosecondToMBPerSec * decompressedBuffer.length / (double) durationMicroseconds);
-      if (decompressedBuffer.length == fullChunkSize) {
+    ByteBuf decompressedBuffer = null;
+    int decompressedBufferSize;
+    ByteBuf newCompressedBuffer = combineBuffer(compressedBuffer, outputDirectBuffer, decompressor.requireMatchingBufferType());
+    try {
+      // Allocate decompression buffer.
+      ByteBuffer compressedByteBuffer = newCompressedBuffer.nioBuffer();
+      decompressedBufferSize = decompressor.getDecompressBufferSize(compressedByteBuffer);
+      decompressedBuffer = outputDirectBuffer ? byteBufAllocator.directBuffer(decompressedBufferSize) :
+          byteBufAllocator.heapBuffer(decompressedBufferSize);
+
+      // Buffer allocated, run decompression.
+      ByteBuffer decompressedByteBuffer = decompressedBuffer.nioBuffer(decompressedBuffer.readerIndex(), decompressedBufferSize);
+      long startTime = System.nanoTime();
+      decompressor.decompress(compressedByteBuffer, decompressedByteBuffer);
+      decompressedBuffer.writerIndex(decompressedBuffer.writerIndex() + decompressedBufferSize);
```

Comment:
Doesn't the decompress method advance this index when writing to compressedBuf ?

Reply:
decompress() takes a ByteBuffer, not ByteBuf, so it does not advance ByteBuf index.